File: src/main/java/moe/giga/discord/CommandHandler.java
package moe.giga.discord;

import moe.giga.discord.annotations.CommandInfo;
import moe.giga.discord.commands.Command;
import moe.giga.discord.contexts.MessageContext;
import moe.giga.discord.contexts.ServerContext;
import moe.giga.discord.contexts.UserContext;
import net.dv8tion.jda.core.JDABuilder;
import net.dv8tion.jda.core.entities.Message;
import net.dv8tion.jda.core.events.message.MessageReceivedEvent;
import net.dv8tion.jda.core.hooks.ListenerAdapter;
import net.dv8tion.jda.core.hooks.SubscribeEvent;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class CommandHandler {
    private final Map<String, Command> commandMap;

    CommandHandler(JDABuilder builder, List<Command> commands) {
        this.commandMap = new HashMap<>();
        for (Command command : commands) {
            CommandInfo info = command.getClass().getAnnotation(CommandInfo.class);
            command.init(builder);
            commandMap.putIfAbsent(info.name(), command);
            for (String alias : info.aliases()) {
                commandMap.putIfAbsent(alias, command);
            }

        }
    }

    private boolean containsCommand(ServerContext SC, Message message) {
        String command = commandArgs(message)[0];
        return command.startsWith(SC.getPrefix()) && commandMap.containsKey(command.substring(1));
    }

    private String[] commandArgs(Message message) {
        return commandArgs(message.getContentRaw());
    }

    private String[] commandArgs(String string) {
        return string.split(" ");
    }

    @SubscribeEvent
    public void handleMessage(MessageReceivedEvent event) {
        ServerContext SC = new ServerContext(event.getGuild());
        if (containsCommand(SC, event.getMessage())) {
            String[] args = commandArgs(event.getMessage());
            Command command = commandMap.get(args[0].substring(1));
            CommandInfo info = command.getClass().getAnnotation(CommandInfo.class);

            if (!event.getAuthor().isBot() || info.allowBots()) {
                MessageContext MC = new MessageContext(event);
                MC.setUC(new UserContext(event.getAuthor()));
                MC.setSC(SC);

                args = Arrays.copyOfRange(args, 1, args.length);
                command.onCommand(MC, args);
            }
        }
    }
}

File: src/main/kotlin/moe/giga/discord/CommandHandler.kt
package moe.giga.discord

import moe.giga.discord.commands.Command
import moe.giga.discord.contexts.MessageContext
import moe.giga.discord.contexts.ServerContext
import net.dv8tion.jda.core.JDABuilder
import net.dv8tion.jda.core.entities.Message
import net.dv8tion.jda.core.events.message.MessageReceivedEvent
import net.dv8tion.jda.core.hooks.SubscribeEvent

class CommandHandler internal constructor(builder: JDABuilder, commands: List<Command>) {
    private val commandMap: Map<String, Command> = commands.associateBy { it.name }
    private val aliasMap: MutableMap<String, String> = HashMap()

    init {
        for (command in commands) {
            command.init(builder)

            // TODO: find a nicer way to do this
            if (command.aliases.isNotEmpty())
                command.aliases.forEach { aliasMap.putIfAbsent(it, command.name) }
        }
    }

    private fun containsCommand(SC: ServerContext, message: Message): Boolean {
        val command = commandArgs(message).first()
        return command.startsWith(SC.prefix) && commandMap.containsKey(command.substring(1))
    }

    private fun commandArgs(message: Message): List<String> {
        return commandArgs(message.contentRaw)
    }

    private fun commandArgs(string: String): List<String> {
        return string.split(" ")
    }

    private fun resolveCommand(lookupString: String): Command? {
        val commandName = aliasMap.getOrDefault(lookupString, lookupString)
        return commandMap[commandName]
    }

    // TODO: allow multiple word commands, probably will need a trie structure to find the commands, or regex lol
    @SubscribeEvent
    fun handleMessage(event: MessageReceivedEvent) {
        val serverContext = ServerContext(event.guild)
        if (containsCommand(serverContext, event.message)) {
            val args = commandArgs(event.message)
            val command = resolveCommand(args[0].substring(1))

            if (command != null) {
                if (!event.author.isBot || command.allowBots) {
                    val mc = MessageContext.Builder().event(event).serverContext(serverContext).build()
                    if (mc.userCtx.allowed(command.level)) {
                        command.onCommand(mc, args.drop(1).toTypedArray())
                    } else {
                        mc.sendError("You are not allowed to run `%s`.").queue()
                    }
                }
            }
        }
    }
}
